Blob URLs with fragment from opaque/unique origins cannot be loaded

cdumez opened the pull request:
#### 2424b6bdaf38c9dd1dd99a2f86a6f11f08561da2
<pre>
Blob URLs with fragment from opaque/unique origins cannot be loaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=243996">https://bugs.webkit.org/show_bug.cgi?id=243996</a>

Reviewed by Darin Adler.

To keep track of the opaque/unique security origins from blob URLs, we rely
on an originMap() in ThreadableBlobRegistry.cpp. This is needed because the
origin is serialized as &quot;null&quot; in the blob URL.

However, unlike the blob storage which always strips any fragment identifier
from Blob URLs because add/removing/querying, we were failing to do so when
using the originMap(). This patch fixes that.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/sandboxed-iframe-expected.txt:
* Source/WTF/wtf/HashMap.h:
(WTF::Y&gt;::remove):
* Source/WTF/wtf/text/StringHash.h:
(WTF::StringViewHashTranslator::translate):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::registerBlobURL):
(WebCore::ThreadableBlobRegistry::unregisterBlobURL):
(WebCore::ThreadableBlobRegistry::getCachedOrigin):

Canonical link: <a href="https://commits.webkit.org/253498@main">https://commits.webkit.org/253498@main</a>
</pre>

darinadler commented on Source/WTF/wtf/text/StringHash.h:
```diff
@@ -237,6 +237,12 @@ namespace WTF {
+
+        static void translate(String& location, StringView view, unsigned hash)
+        {
+            location = view.toString();
```
A little sad that when the `StringView` covers an entire `String`, we will construct a new `String` and can’t use the `String` it was made from. We could fix this by introducing a type that contains both a `StringView` and a reference to an underlying `String`, but we need not do that in this patch. It might be tricky to design this so it’s not cumbersome or risky to use.

cdumez committed:
Blob URLs with fragment from opaque/unique origins cannot be loaded
https://bugs.webkit.org/show_bug.cgi?id=243996

Reviewed by Darin Adler.

To keep track of the opaque/unique security origins from blob URLs, we rely
on an originMap() in ThreadableBlobRegistry.cpp. This is needed because the
origin is serialized as "null" in the blob URL.

However, unlike the blob storage which always strips any fragment identifier
from Blob URLs because add/removing/querying, we were failing to do so when
using the originMap(). This patch fixes that.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/sandboxed-iframe-expected.txt:
* Source/WTF/wtf/HashMap.h:
(WTF::Y>::remove):
* Source/WTF/wtf/text/StringHash.h:
(WTF::StringViewHashTranslator::translate):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::registerBlobURL):
(WebCore::ThreadableBlobRegistry::unregisterBlobURL):
(WebCore::ThreadableBlobRegistry::getCachedOrigin):

Canonical link: https://commits.webkit.org/253498@main
